### PR TITLE
Package ptr with generics

### PR DESCRIPTION
**Pull request by lzap**
Up until now we used the `ptr` package from smithy library. With Go 1.18, we can now take advantage of generics and have a single function `ptr.To` that will work with any type. This patch creates a new `internal/ptr` package with functions `To` and `From` for easy pointer copying. On top of that, there are also `ToInt64` (and 32) variants so you don't have to type `ptr.To(int64(42))` for constants because the default type for number is `int`.

Feel free to add new functions for other types (unsigned ints) or floats if needed: https://go.dev/tour/basics/11

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
our `fmt` still formats for go 1.16 I guess :)
TBH I do not see a big value in this usage of generic, not much less typing here, I mean it's great example of how to use generics, but the usage seems mostly the same as we use only String and int, where we still endup using different functions per type xD

**Comment by lzap** [comment]
I understand that it does not look obvious when you see the diff, but it feels different. You simply type `ptr.To` and don't think about a type. Well, unless it's a integer constant, but that does not make it worse than it was before.

So I could not get the current GHA action working, but I realized that actually the golangci-lint does support both fmt and imports, so I enabled it and let's just use a single tool for everything.

**Comment by lzap** [comment]
Since I renamed GHA job names, the status won't perhaps be reported at all. Or I am not sure why it's stuck.

File: .golangci.yml [hunk 1]
```diff
@@ -10,15 +10,22 @@ linters:
...
     - scopelint # deprecated by exportloopref
+    - gci # too strict
+    - gofumpt # too strict
+    - contextcheck # too restrictive
 linters-settings:
   govet:
     check-shadowing: true
+  gosec:
+    excludes:
+      - G112 # slowloris attack - TODO (easyfix)
```
Comment (lzap) [inline .golangci.yml:28]: This is easy fix, will do in a different PR. Just two occurrences but new setting is needed.

File: .golangci.yml [hunk 2]
```diff
@@ -10,15 +10,22 @@ linters:
     - performance
     - sql
     - module
-    - unused
+    - format
+    - import
```
Comment (lzap) [inline .golangci.yml:14]: These two (format, import) actually enables "go fmt" and "goimports" chekcing

The unused was accidentally added, I thought it is some kind of "unused code" check while it is actually a deprecated group.

**Comment by ezr-ondrej** [comment]
> You simply type `ptr.To` and don't think about a type. Well, unless it's a integer constant, but that does not make it worse than it was before.

I understand this, but we are using only string and int at the moment and I see us only using those, because for other types, you usually have a variable and use dereference, so you'll always need to think about type, because string is `To` and integer is `ToIntXX` so the point of this code seems really to showcase usage of generics, which I love and hate from `C++` where after its introduction it looked cool as hell, but it made everything more confusing (especially debugging).

I'm not against this, I'm just saying we are now maintaining a code, that feels like it should be a shared go library (similar to `smithy-go/ptr`)

**Review by ezr-ondrej — APPROVED** [review]
I love the GH action improvements!
And I'm all for getting rid of unecessary dependencies and the code looks quite clean :+1:

**Comment by lzap** [comment]
>  you'll always need to think about type

This statement is not correct. You only need to think about it, if you are making a pointer to local variable which you created via type inference which defaults to `int` or `float64`. It works fine for functions with variables passed in as arguments:

```go
func Function(x int64) {
 _ = ptr.To(x)
}
```

Granted we do have a bunch of them but mostly in tests/stubs, if you look closer to `RunInstances` function you will see that it actually gets easier and cleaner. Fun fact: I actually skipped these because I searched only for `ptr.` but this code actually uses `aws.Int32` which is a different package from AWS SDK to do the same thing. Let me fix that as well.

**Comment by lzap** [comment]
I just pushed a commit that also changes the real EC2 client, not stubs/tests. This is where I think the new `To` and `From` functions really shine. :-)

File: internal/clients/http/ec2/ec2_client.go [hunk 3]
```diff
@@ -167,8 +167,8 @@ func (c *ec2Client) RunInstances(ctx context.Context, name *string, amount int32
 				ResourceType: types.ResourceTypeInstance,
 				Tags: []types.Tag{
 					{
-						Key:   aws.String("Name"),
-						Value: aws.String(*name),
+						Key:   ptr.To("Name"),
+						Value: ptr.To(*name),
```
Comment (ezr-ondrej) [inline internal/clients/http/ec2/ec2_client.go:171]: Interesting code actually :thinking: :smile:

**Review by ezr-ondrej — APPROVED** [review]
Thanks @lzap ! I like the showcase of generics, lets get some hands-on with it :tada:

**Comment by ezr-ondrej** [comment]
Thanks @lzap ! :+1:
